File: python/endpoints/docking/export_docks.py
from clearpath_config.common.utils.yaml import read_yaml
import rclpy
from rclpy.node import Node
from clearpath_dock_msgs.srv import ExportData

ROBOT_CONFIG_PATH = '/etc/clearpath/robot.yaml'
LOGGING_NAME = 'ExportDocks'


class DeleteDock(Node):

    def __init__(self, namespace: str):
        super().__init__('export_docks_client')
        self._export_docks_client = self.create_client(ExportData, f'/{namespace}/docking/dock_manager/export_docks')

        while not self._export_docks_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().info(f'[{LOGGING_NAME}] /{namespace}/docking/dock_manager/export_docks service not available, waiting again...')

    def export_docks(self):
        req = ExportData.Request()
        self.future = self._export_docks_client.call_async(req)
        self.future.add_done_callback(self.export_docks_srv_response_cb)

    def export_docks_srv_response_cb(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'[{LOGGING_NAME}] Docks exported successfully!')
            else:
                self.get_logger().error(f'[{LOGGING_NAME}] Failed to export docks: {response.message}')
        except Exception as e:
            self.get_logger().error(f'[{LOGGING_NAME}] Service call failed: {e}')

        self.destroy_node()
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    robot_config = read_yaml(ROBOT_CONFIG_PATH)
    namespace = robot_config['system']['ros2']['namespace']

    export_docks_client = DeleteDock(namespace)
    export_docks_client.export_docks()

    rclpy.spin(export_docks_client)


if __name__ == '__main__':
    main()

File: cpp/endpoints/docking/export_docks.cpp
#include "rclcpp/rclcpp.hpp"
#include "clearpath_dock_msgs/srv/export_data.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>
#include <yaml-cpp/yaml.h>
  
using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  YAML::Node robot_config = YAML::LoadFile("/etc/clearpath/robot.yaml");
  std::string namespace_ = robot_config["system"]["ros2"]["namespace"].as<std::string>();

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("export_docks_client");
  rclcpp::Client<clearpath_dock_msgs::srv::ExportData>::SharedPtr client =
    node->create_client<clearpath_dock_msgs::srv::ExportData>("/" + namespace_ + "/" + std::string("docking/dock_manager/export"));

  auto request = std::make_shared<clearpath_dock_msgs::srv::ExportData::Request>();

  while (!client->wait_for_service(2s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("delete_dock_client"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("export_docks_client"), "/%s/docking/dock_manager/export service not available, waiting again...", namespace_.c_str());
  }

  auto result = client->async_send_request(request);

  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    if (result.get()->success)
    {
      RCLCPP_INFO(rclcpp::get_logger("export_docks_client"), "Docks exported successfully!");
    }
    else
    {
      RCLCPP_ERROR(rclcpp::get_logger("export_docks_client"), "Failed to export docks!");
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("export_docks_client"), "Failed to call service /%s/docking/dock_manager/export", namespace_.c_str());
  }

  rclcpp::shutdown();
  return 0;
}
